File: mobile/lib/features/surgeon_gate/domain/gate_state.dart
/// Mirrors `aso_host::domain::GateState`. The all-four rule lives in the Rust
/// core; this type lets the UI render what is outstanding without inventing a
/// second definition of "affirmed".
enum GateAffirmationKind {
  policy('policy', 'Controlling policy',
      'This is the policy and version that governs this request on the date of service.'),
  section('section', 'Criterion section',
      'This is the section of that policy the request must satisfy.'),
  pathway('pathway', 'Surgical pathway',
      'This is the operation I intend to perform.'),
  plan('plan', 'Operative plan',
      'The described levels, approach and extent match my operative plan.');

  const GateAffirmationKind(this.wire, this.label, this.prompt);

  final String wire;
  final String label;
  final String prompt;

  static GateAffirmationKind fromWire(String v) =>
      values.firstWhere((k) => k.wire == v);
}

class GateState {
  const GateState({required this.affirmed, required this.outstanding});

  final bool affirmed;
  final List<GateAffirmationKind> outstanding;

  factory GateState.fromJson(Map<String, dynamic> json) => GateState(
        affirmed: json['affirmed'] as bool,
        outstanding: (json['outstanding'] as List<dynamic>)
            .map((v) => GateAffirmationKind.fromWire(v as String))
            .toList(),
      );
}

File: web/src/features/surgeon-gate/model/gate-state.ts
// Wire types mirror aso_host::affirmation; GateState adapts the committed
// snapshot for the existing affirmation list.

export const GATE_KINDS = ['policy', 'section', 'pathway', 'plan'] as const;
export type GateAffirmationKind = (typeof GATE_KINDS)[number];

export interface GateState {
  affirmed: boolean;
  outstanding: GateAffirmationKind[];
}

export interface GateSnapshot {
  caseId: string;
  affirmed: GateAffirmationKind[];
  gateAffirmedAt: string | null;
  gateAffirmedBy: string | null;
}

export interface GateMutation {
  commandId: string;
  kind: GateAffirmationKind;
}

export interface GateCommandResult extends GateMutation {
  caseId: string;
  action: 'affirm' | 'remove';
  gate: GateSnapshot;
  committedAt: string;
}

export function gateStateFromSnapshot(snapshot: GateSnapshot): GateState {
  const outstanding = GATE_KINDS.filter((kind) => !snapshot.affirmed.includes(kind));
  return { affirmed: outstanding.length === 0, outstanding };
}

export const gateKindLabel: Record<GateAffirmationKind, string> = {
  policy: 'Controlling policy',
  section: 'Criterion section',
  pathway: 'Surgical pathway',
  plan: 'Operative plan',
};

export const gateKindPrompt: Record<GateAffirmationKind, string> = {
  policy: 'This is the policy and version that governs this request on the date of service.',
  section: 'This is the section of that policy the request must satisfy.',
  pathway: 'This is the operation I intend to perform.',
  plan: 'The described levels, approach and extent match my operative plan.',
};
